### PR TITLE
feat(security/pipeline): Add Pipeline to add more possibilities to the security builder

### LINKED ISSUE #9: Security Factory: Add Built-In for 42 school

### LINKED ISSUE #10: Security Factory: Add Built-In for Gitlab

### LINKED ISSUE #11: Security Factory: Add Built-In for Github

### LINKED ISSUE #21: Security Factory: compareWithGeneratedHMAC
Add a factory that allows to compare 2 inputs with a generated HMAC. 

Useful for Github or Twitch webhooks

Example of config (not final)
```yaml
  security:
  - getHeader:
      name: X-Hook-Secret
  - compareWithGeneratedHMAC:
      secret: '{{ .env.HOOK_SECRET }}'
      algorithm: sha256
```

### LINKED ISSUE #22: Security Factory: compareWithAllowedIPs / compareWithBlockedIPs
Add a factory that allows to compare an IP via a blocklist or an allowlist

Example of config (not final)
```yaml
  security:
  - getClientIP:
  - compareWithAllowedIPs:
      ips: ['127.0.0.1']
# OR
  security:
  - getClientIP:
  - compareWithBlockedIPs: 
      ips: ['127.0.0.1']
```

### PR DESCRIPTION
**Pull request by 42Atomys**
**Relative Issues:** Resolve #22 , resolve #21, resolve #9, resolve #10, resolve #11

**Describe the pull request**
The current version of the security factories works in a very basic way and does not allow advanced uses like comparing several objects of different steps or allowing later the comparison of HMAC which is for example useful for the webhook of GitHub or Twitch. 

For this we had to rewrite version 1 of the `factory builder` to add a so-called `Pipeline` overlay

> **In other words, stop me, I recoded the GitHub Actions system in one night, f*****

**TODO**
- [x] Build fonctionnal pipeline code
- [x] Integrate already implemented the `header` and `compare` factory in this version
- [x] Structure the code and clean it
- [x] Add tests suites to Pipeline, Factory
- [x] Rewrite configuration tests suites with this new version
- [x] Rewrite handlers tests suites with this new version
- [x] Implements new factory to resolve #22 
- [x] Implements new factory to resolve #21
- [x] Implements new factory to resolve #10
- [x] Implements new factory to resolve #11

**Checklist**

- [x] I have linked the relative issue to this pull request
- [x] I have made the modifications or added tests related to my PR
- [x] I have added/updated the documentation for my RP
- [x] I put my PR in Ready for Review only when all the checklist is checked

**Breaking changes ?**
yes

**Additional context**

Concrete example, actual static comparaison of an header with `factory/v1.Builder`
```yaml
security:
  - getHeader:
      name: X-Hook-Secret
  - compareWithStaticValue:
      values: ['foo', 'bar']
  ```` 

New version builder with `factory/v2.Pipeline`
```yaml
security:
- header:
    id: secretHeader
    inputs:
    - name: headerName
      valueFrom:
        envRef: HEADER
- compare:
    id: secretCompare
    inputs:
    - name: first
      value: '{{ .Outputs.secretHeader.value }}'
    - name: second
      value: 'developmentSecret'
      valueFrom:
        envRef: HOOK_SECRET
```

### CODE REVIEW

**Comment by 42Atomys** [comment]
Config to allow IP based on `X-Forwarded-For` header:
```yaml
  security:
  - header:
      inputs:
      - name: headerName
        value: X-Forwarded-For
  - hasPrefix:
      inputs:
      - name: text
        value: '{{ .Outputs.header.value }}'
      - name: prefix
        values: ['127.0.0.1', '[::1]']
```
Config to allow IP based on RemoteAddr
```yaml
  security:
  - hasPrefix:
      inputs:
      - name: text
        value: '{{ .Inputs.request.RemoteAddr }}'
      - name: prefix
        values: ['127.0.0.1', '[::1]']
```

Config to reject (only add `inverse: true`)
```yaml
  security:
  - hasPrefix:
      inverse: true
      inputs:
      - name: text
        value: '{{ .Inputs.request.RemoteAddr }}'
      - name: prefix
        values: ['127.0.0.1', '[::1]']
```

**Comment by 42Atomys** [comment]
Webhook configuration for Github (Tested OK | Security OK)

```yaml
- name: github
  entrypointUrl: /webhooks/github
  security:
  - generate_hmac_256:
      id: signature
      inputs:
      - name: payload
        value: '{{ .Inputs.payload }}'
      - name: secret
        value: test
  - header:
      id: headerSignature
      inputs:
      - name: headerName
        value: X-Hub-Signature-256
  - compare:
      inputs:
      - name: first
        value: '{{ .Outputs.headerSignature.value }}'
      - name: second
        value: 'sha256={{ .Outputs.signature.value }}'
  storage:
  - type: redis
    specs:
      host: localhost
      port: '6379'
      database: 0
      key: github
 ```

**Comment by 42Atomys** [comment]
Webhook configuration for Gitlab (Tested OK | Security OK)
```yaml
- name: gitlab
  entrypointUrl: /webhooks/gitlab
  security:
  - header:
      inputs:
      - name: headerName
        value: X-Gitlab-Token
  - compare:
      inputs:
      - name: first
        value: '{{ .Outputs.header.value }}'
      - name: second
        valueFrom:
          envRef: GITLAB_WEBHOOK_TOKEN
  storage:
  - type: redis
    specs:
      host: localhost
      port: '6379'
      database: 0
      key: gitlab
```

File: internal/config/configuration.go [hunk 1]
```diff
@@ -7,7 +7,7 @@ import (
 	"github.com/rs/zerolog/log"
 	"github.com/spf13/viper"
 
-	"42stellar.org/webhooks/pkg/factory"
+	factory "42stellar.org/webhooks/pkg/factory"
```
Comment (rgaiffe) [inline internal/config/configuration.go:10]: Dont need rename factory per factory :smart:

File: pkg/factory/registry.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+var (
+	// FunctionMap contains the map of function names to their respective functions
+	// This is used to validate the function name and to get the function by name
+	factoryMap = map[string]*Factory{
+		"debug":             newFactory(&debugFactory{}),
+		"header":            newFactory(&headerFactory{}),
+		"compare":           newFactory(&compareFactory{}),
+		"hasPrefix":         newFactory(&hasPrefixFactory{}),
+		"hasSuffix":         newFactory(&hasSuffixFactory{}),
+		"generate_hmac_256": newFactory(&generateHMAC256Factory{}),
```
Comment (rgaiffe) [inline pkg/factory/registry.go:16]: Camel case

File: pkg/factory/f_has_prefix.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		{false, reflect.TypeOf(&InputConfig{}), "prefix", &InputConfig{}},
+	}
+}
+
+func (*hasPrefixFactory) DefinedOutputs() []*Var {
+	return []*Var{
+		{false, reflect.TypeOf(false), "result", false},
+	}
+}
+
```
Comment (rgaiffe) [inline pkg/factory/f_has_prefix.go:27]: Commentary

**Review by rgaiffe — CHANGES_REQUESTED** [review]
LGTM with few changes.

**Review by rgaiffe — APPROVED** [review]

**Pull request merged**